File: HW2/Q6/benchmark/analyze.py
#!/usr/bin/env python3
"""
analyze.py -- turn benchmark/results.csv into the tables and plots the report
needs.

Produces on stdout, in the layout of report_format.md:
  * runtime table   (median t_algo, seconds)
  * speed-up table  S(P) = T1 / TP
  * efficiency table E(P) = S(P) / P
  * communication vs computation table (% of t_algo spent in MPI_Allreduce)
  * convergence round counts

And writes to benchmark/plots/:
  speedup.png, efficiency.png, comm_fraction.png

The reported time per configuration is the MEDIAN over trials: it is robust
against a single slow run caused by another job sharing the node.

T1 is the MPI binary at P=1, not the sequential program, so the speed-up
measures parallel scaling of one implementation rather than a comparison of
two different programs.

Usage: python3 benchmark/analyze.py [results.csv]
"""

import csv
import os
import statistics
import sys
from collections import defaultdict

HERE = os.path.dirname(os.path.abspath(__file__))
CSV = sys.argv[1] if len(sys.argv) > 1 else os.path.join(HERE, "results.csv")
PLOTS = os.path.join(HERE, "plots")


def load(path):
    if not os.path.exists(path):
        sys.exit(f"no results file at {path}\nRun benchmark/run_bench.sh first.")
    rows = []
    with open(path) as f:
        for r in csv.DictReader(f):
            try:
                r["P"] = int(r["P"])
                r["rounds"] = int(r["rounds"])
                for k in ("t_input", "t_compute", "t_comm", "t_total"):
                    r[k] = float(r[k])
                r["t_algo"] = r["t_compute"] + r["t_comm"]
            except (ValueError, KeyError):
                continue
            rows.append(r)
    if not rows:
        sys.exit("results file contains no usable rows")
    return rows


def aggregate(rows):
    """(dataset, P) -> median of each metric across trials."""
    buckets = defaultdict(lambda: defaultdict(list))
    for r in rows:
        key = (r["dataset"], r["P"])
        for m in ("t_input", "t_compute", "t_comm", "t_total", "t_algo"):
            buckets[key][m].append(r[m])
        buckets[key]["rounds"].append(r["rounds"])
    out = {}
    for key, met in buckets.items():
        out[key] = {m: statistics.median(v) for m, v in met.items()}
        out[key]["trials"] = len(met["t_algo"])
    return out


def table(title, header, body):
    print(f"\n### {title}\n")
    print("| " + " | ".join(header) + " |")
    print("|" + "|".join([":--"] + [":-:"] * (len(header) - 1)) + "|")
    for row in body:
        print("| " + " | ".join(row) + " |")


def main():
    rows = load(CSV)
    agg = aggregate(rows)

    datasets = []
    for r in rows:
        if r["dataset"] not in datasets:
            datasets.append(r["dataset"])
    procs = sorted({r["P"] for r in rows})

    print("# Q6 Benchmark Analysis")
    print(f"\nSource: `{os.path.relpath(CSV)}`  |  process counts: "
          f"{', '.join(map(str, procs))}  |  statistic: median over trials")
    print("\n`t_algo` = `t_compute` + `t_comm` (excludes file reading and "
          "distribution).")

    # ---- runtime -----------------------------------------------------------
    body = []
    for d in datasets:
        cells = [d]
        for p in procs:
            a = agg.get((d, p))
            cells.append(f"{a['t_algo']:.4f}" if a else "--")
        body.append(cells)
    table("Runtime — median `t_algo` (seconds)",
          ["Input"] + [f"P={p}" for p in procs], body)

    # ---- speed-up ----------------------------------------------------------
    speed = {}
    body = []
    for d in datasets:
        base = agg.get((d, 1))
        cells = [d]
        for p in procs:
            a = agg.get((d, p))
            if a and base and a["t_algo"] > 0:
                s = base["t_algo"] / a["t_algo"]
                speed[(d, p)] = s
                cells.append(f"{s:.2f}")
            else:
                cells.append("--")
        body.append(cells)
    table("Speed-up  S(P) = T1 / TP",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- efficiency --------------------------------------------------------
    body = []
    for d in datasets:
        cells = [d]
        for p in procs:
            s = speed.get((d, p))
            cells.append(f"{100.0 * s / p:.1f}%" if s else "--")
        body.append(cells)
    table("Efficiency  E(P) = S(P) / P",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- communication share ----------------------------------------------
    body = []
    for d in datasets:
        cells = [d]
        for p in procs:
            a = agg.get((d, p))
            if a and a["t_algo"] > 0:
                cells.append(f"{100.0 * a['t_comm'] / a['t_algo']:.1f}%")
            else:
                cells.append("--")
        body.append(cells)
    table("Communication share — `t_comm` as % of `t_algo`",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- rounds ------------------------------------------------------------
    body = []
    for d in datasets:
        cells = [d]
        for p in procs:
            a = agg.get((d, p))
            cells.append(f"{int(a['rounds'])}" if a else "--")
        body.append(cells)
    table("Convergence rounds (deterministic, machine-independent)",
          ["Input size"] + [f"P={p}" for p in procs], body)

    # ---- plots -------------------------------------------------------------
    try:
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
    except ImportError:
        print("\n(matplotlib not available - plots skipped)")
        return

    os.makedirs(PLOTS, exist_ok=True)

    def lineplot(fname, ylabel, title, valfn, ideal=False):
        fig, ax = plt.subplots(figsize=(7, 4.5))
        for d in datasets:
            xs, ys = [], []
            for p in procs:
                v = valfn(d, p)
                if v is not None:
                    xs.append(p); ys.append(v)
            if xs:
                ax.plot(xs, ys, marker="o", label=d)
        if ideal:
            ax.plot(procs, procs, "k--", alpha=0.4, label="ideal (linear)")
        ax.set_xlabel("Number of MPI processes (P)")
        ax.set_ylabel(ylabel)
        ax.set_title(title)
        ax.set_xticks(procs)
        ax.grid(alpha=0.3)
        ax.legend(fontsize=8)
        fig.tight_layout()
        path = os.path.join(PLOTS, fname)
        fig.savefig(path, dpi=150)
        plt.close(fig)
        return path

    p1 = lineplot("speedup.png", "Speed-up  S(P)",
                  "Q6 Connected Components — Speed-up",
                  lambda d, p: speed.get((d, p)), ideal=True)
    p2 = lineplot("efficiency.png", "Efficiency  E(P)",
                  "Q6 Connected Components — Efficiency",
                  lambda d, p: (speed[(d, p)] / p) if (d, p) in speed else None)
    p3 = lineplot("comm_fraction.png", "t_comm / t_algo",
                  "Q6 — Communication share of algorithm time",
                  lambda d, p: (agg[(d, p)]["t_comm"] / agg[(d, p)]["t_algo"])
                  if (d, p) in agg and agg[(d, p)]["t_algo"] > 0 else None)

    print("\n### Plots written\n")
    for p in (p1, p2, p3):
        print(f"- `{os.path.relpath(p)}`")


if __name__ == "__main__":
    main()

File: Q6/tools/analyze.cpp
// analyze.cpp
// ---------------------------------------------------------------------------
// Q6 analysis tool (C++ replacement for the Python script).
//
// Reads benchmark/results.csv and writes, on stdout, the tables the report
// format asks for: runtime, speed-up, efficiency, communication share, and
// convergence round counts. Writes three SVG charts into benchmark/plots/.
//
// SVG is generated directly (see harness.hpp) so no plotting library is
// needed; it embeds in a report as is, and converts to PNG if a raster is
// required.
//
// The value reported per configuration is the MEDIAN over trials. T1 is the
// MPI binary at P=1, not the sequential program, so the speed-up measures
// parallel scaling of one implementation.
//
// Build: g++ -O2 -std=c++17 -o analyze analyze.cpp
// Run  : ./tools/analyze [results.csv]     (from the Q6 project root)
// ---------------------------------------------------------------------------

#include <cstdio>
#include <map>
#include <string>
#include <vector>

#include "harness.hpp"

using harness::fmt;
using harness::median;
using harness::Series;

struct Key {
    std::string ds;
    int         P;
    bool operator<(const Key& o) const {
        return ds != o.ds ? ds < o.ds : P < o.P;
    }
};

struct Agg { double input = 0, compute = 0, comm = 0, total = 0, algo = 0, rounds = 0; };

int main(int argc, char** argv) {
    const std::string path = (argc > 1) ? argv[1] : "benchmark/results.csv";

    auto rows = harness::readCsv(path);
    if (rows.empty()) {
        std::fprintf(stderr,
                     "no usable results at %s\nRun ./tools/run_bench first "
                     "(from the Q6 project root).\n", path.c_str());
        return 1;
    }

    std::map<Key, std::vector<double>> vIn, vComp, vComm, vTot, vRounds;
    std::vector<std::string> dsets;
    std::vector<int> procs;

    for (auto& r : rows) {
        Key k{r["dataset"], std::atoi(r["P"].c_str())};
        vIn[k].push_back(std::atof(r["t_input"].c_str()));
        vComp[k].push_back(std::atof(r["t_compute"].c_str()));
        vComm[k].push_back(std::atof(r["t_comm"].c_str()));
        vTot[k].push_back(std::atof(r["t_total"].c_str()));
        vRounds[k].push_back(std::atof(r["rounds"].c_str()));
        if (std::find(dsets.begin(), dsets.end(), k.ds) == dsets.end())
            dsets.push_back(k.ds);
        if (std::find(procs.begin(), procs.end(), k.P) == procs.end())
            procs.push_back(k.P);
    }
    std::sort(procs.begin(), procs.end());

    std::map<Key, Agg> agg;
    for (auto& kv : vComp) {
        const Key& k = kv.first;
        Agg a;
        a.input   = median(vIn[k]);
        a.compute = median(vComp[k]);
        a.comm    = median(vComm[k]);
        a.total   = median(vTot[k]);
        a.rounds  = median(vRounds[k]);
        a.algo    = a.compute + a.comm;
        agg[k] = a;
    }
    auto get = [&](const std::string& d, int p, Agg& out) {
        auto it = agg.find(Key{d, p});
        if (it == agg.end()) return false;
        out = it->second;
        return true;
    };

    std::printf("# Q6 Benchmark Analysis - Connected Components\n");
    std::printf("\nSource: `%s`  |  process counts:", path.c_str());
    for (int p : procs) std::printf(" %d", p);
    std::printf("  |  statistic: median over trials\n");
    std::printf("\n`t_algo` = `t_compute` + `t_comm` (excludes file reading and "
                "distribution).\n");

    std::vector<std::string> head{"Input size"};
    for (int p : procs) head.push_back("P=" + std::to_string(p));

    {
        std::vector<std::vector<std::string>> body;
        for (auto& d : dsets) {
            std::vector<std::string> row{d};
            for (int p : procs) { Agg a; row.push_back(get(d, p, a) ? fmt(a.algo, 4) : "--"); }
            body.push_back(row);
        }
        harness::printTable("Runtime - median `t_algo` (seconds)", head, body);
    }

    std::map<Key, double> speed;
    {
        std::vector<std::vector<std::string>> body;
        for (auto& d : dsets) {
            std::vector<std::string> row{d};
            Agg base; const bool haveBase = get(d, 1, base);
            for (int p : procs) {
                Agg a;
                if (haveBase && get(d, p, a) && a.algo > 0) {
                    const double v = base.algo / a.algo;
                    speed[Key{d, p}] = v;
                    row.push_back(fmt(v, 2));
                } else row.push_back("--");
            }
            body.push_back(row);
        }
        harness::printTable("Speed-up  S(P) = T1 / TP", head, body);
    }

    {
        std::vector<std::vector<std::string>> body;
        for (auto& d : dsets) {
            std::vector<std::string> row{d};
            for (int p : procs) {
                auto it = speed.find(Key{d, p});
                row.push_back(it == speed.end() ? "--" : fmt(100.0 * it->second / p, 1) + "%");
            }
            body.push_back(row);
        }
        harness::printTable("Efficiency  E(P) = S(P) / P", head, body);
    }

    {
        std::vector<std::vector<std::string>> body;
        for (auto& d : dsets) {
            std::vector<std::string> row{d};
            for (int p : procs) {
                Agg a;
                row.push_back((get(d, p, a) && a.algo > 0)
                                  ? fmt(100.0 * a.comm / a.algo, 1) + "%" : "--");
            }
            body.push_back(row);
        }
        harness::printTable("Communication share - `t_comm` as % of `t_algo`",
                            head, body,
                            "The Allreduce moves V ints per round regardless of "
                            "P, while the per-rank edge work falls as 1/P.");
    }

    {
        std::vector<std::vector<std::string>> body;
        for (auto& d : dsets) {
            std::vector<std::string> row{d};
            for (int p : procs) {
                Agg a;
                row.push_back(get(d, p, a) ? fmt(a.rounds, 0) : "--");
            }
            body.push_back(row);
        }
        harness::printTable("Convergence rounds", head, body,
                            "Deterministic and machine-independent: these "
                            "depend on the graph and the partition, not on "
                            "timing.");
    }

    std::system("mkdir -p benchmark/plots");

    auto seriesOf = [&](bool useSpeed, int mode) {
        // mode 0 = algo time, 1 = comm fraction, 2 = efficiency
        std::vector<Series> out;
        for (auto& d : dsets) {
            Series s; s.name = d;
            for (int p : procs) {
                if (useSpeed) {
                    auto it = speed.find(Key{d, p});
                    if (it == speed.end()) continue;
                    s.x.push_back(p);
                    s.y.push_back(mode == 2 ? it->second / p : it->second);
                } else {
                    Agg a;
                    if (!get(d, p, a)) continue;
                    s.x.push_back(p);
                    s.y.push_back(mode == 1 ? (a.algo > 0 ? a.comm / a.algo : 0.0)
                                            : a.algo);
                }
            }
            if (!s.x.empty()) out.push_back(s);
        }
        return out;
    };

    harness::svgPlot("benchmark/plots/time.svg", "Q6 - Execution time vs processes",
                     "Number of MPI processes (P)", "t_algo (s, log scale)",
                     seriesOf(false, 0), true, false);
    harness::svgPlot("benchmark/plots/speedup.svg", "Q6 - Speed-up vs processes",
                     "Number of MPI processes (P)", "Speed-up  S(P)",
                     seriesOf(true, 0), false, true);
    harness::svgPlot("benchmark/plots/efficiency.svg", "Q6 - Efficiency vs processes",
                     "Number of MPI processes (P)", "Efficiency  E(P)",
                     seriesOf(true, 2), false, false);
    harness::svgPlot("benchmark/plots/comm_fraction.svg",
                     "Q6 - Communication share of algorithm time",
                     "Number of MPI processes (P)", "t_comm / t_algo",
                     seriesOf(false, 1), false, false);

    std::printf("\n### Plots written\n\n");
    std::printf("- `benchmark/plots/time.svg`\n");
    std::printf("- `benchmark/plots/speedup.svg`\n");
    std::printf("- `benchmark/plots/efficiency.svg`\n");
    std::printf("- `benchmark/plots/comm_fraction.svg`\n");
    return 0;
}
